[Yjs Collab] Reliable sync with the backend

dmonad opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?

This PR implements a reliable sync protocol for collaborative editing. 

## Why?
Previously, Gutenberg implemented a basic [Yjs](https://docs.yjs.dev/) editor binding. Changes were synced using y-webrtc when collaborative editing is enabled. However, the previous implementation did not reliably sync the backend-state with the Yjs document. The collaborative document was not preserved over different sessions, which might lead to content duplication and even loss of content.

Furthermore, even in non-collaborative sessions, it is quite easy to overwrite content from other users. Changes that are created by different users at the same time are not reconciled.

Related discussion on the considerations to improve the current sync approach: https://github.com/WordPress/gutenberg/discussions/65012

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

This PR implements a reliable sync protocol for syncing backend state with a Yjs document.

- We save the Yjs document in an HTML comment `<!-- y:gutenberg .. -->`, which is stored in the HTML document stored in WordPress
- When the client notices that the Yjs document stored in the comment does not reflect the state of the HTML document, we assume that the changes were generated by the backend. We reconcile the changes and incorporate them into the Yjs document. Then we save the rebased Yjs document once again to the `<!-- y:gutenberg -->` comment and send it to the backend.
- The changes from a backend are applied as a simulated "system user". This ensures that everyone applies the changes in the same way, to avoid content duplication.

Furthermore, this PR improves on the general user-experience of collaborative editing.

- When collaborative editing is enabled, the autosave interval is set to 5 seconds.
- The banners mentioning that there is a newer version of a document has been disabled in collaborative editing.
- Don't sync unnecessary attributes like selection state.
- Improved granularity of changes for better conflict resolution.

I documented the sync protocol in greater detail in [`packages/sync/CODE.md`](https://github.com/dmonad/gutenberg/blob/y/packages/sync/CODE.md).

## Testing Instructions

- [Follow the instructions to enable collaborative editing](https://github.com/dmonad/gutenberg/blob/y/packages/sync/CODE.md#enable-the-experiment).
- It should be possible to edit the same post in different browser windows. Changes made in one browser window will eventually show up in the other browser windows.
- Conflicts are resolved automatically.
- It should be possible to modify the HTML content directly. The collaborative clients will reconcile the changes and sync up.

## Things to note

- Syncing through y-webrtc has been disabled by default. After review, I'm concerned about security risks, because the access control (using a shared password) could be circumvented. We may want to improve the approach in a separate PR.

## TODOs

-  [x] add a separate experiments flag to enable y-webrtc with a password.

----

I hope that the sync approach implemented in this PR will be a good foundation for future work on making different parts of Gutenberg more collaborative. If we want to make Gutenberg / WordPress even more collaborative in the future, we need to share knowledge and make collaborative editing a priority for all future work. It is quite hard to add collaborative editing on-top of an existing code-base. Ideally, future work is built with collaborative editing in mind.

In this regard, I want to host a review & demo session on **Monday, January 6 7pm UTC**. This would be a good place to ask questions. Also, we can try to break this implementation and measure the overhead of decreasing the autosave interval.

- https://lu.ma/vd7uhvhw
- Or simply join the Zoom session: `"https://us02web.zoom.us/j/" + "86065095542"`

dmonad committed:
[Yjs Collab] WIP - Sync changes on (auto)save using Yjs

Previously, Gutenberg implemented a basic Yjs editor binding. Changes
were synced using y-webrtc when collaborative editing is enabled.
However, the previous implementation did not reliably sync the
backend-state with the Yjs document. Hence, the previous implementation
could result in content duplication, or loss of changes that were
created in the collaborative session.

This commit implements the initial approach to sync the backend document
(HTML encoded) with the Yjs document without the server understanding
Yjs.

Following things were implemented:

- We save the Yjs document in an HTML comment, that is stored in the
  HTML document stored in WordPress
- When the client notices that the Yjs document stored in the comment
  does not reflect the state of the HTML document, we assume that the
  changes were generated by the backend. Hence we update the document
  that is stored in the HTML comment, and merge the Yjs update to the
  Yjs document that is used in the collaborative session. Then we save
  the merged document once again to the backend.

The changes from a backend are applied as a simulated "system user".
This should guarantee that everyone applies the changes in the same way,
so we avoid content duplication. The
approach in this commit is still insufficient. We need to store
additional information in the backend to ensure that certain edge-cases
are handled.

dmonad committed:
[Yjs Collab] update used dependencies

dmonad committed:
[Yjs Collab] fix collab issues introduced in upstream - ensure that entities are always registered.

dmonad committed:
[Yjs Collab] Use heartbeat api to pull ydoc updates

dmonad committed:
[Yjs Collab] cleanup

dmonad committed:
[Yjs Collab] improve diffing

dmonad committed:
[Yjs Collab] loss of selection & focus when content is saved

dmonad committed:
[Yjs Collab] add origin to changes originating from gutenberg

dmonad committed:
[Yjs Collab] documentation

dmonad committed:
[Yjs Collab] improve docs

szepeviktor commented:
📜 https://beta.yjs.dev/docs/introduction

dmonad committed:
[Yjs Collab] several bugfixes

dmonad committed:
[Yjs Collab] add different feature flags for sync on save / heartbeat sync / webrtc sync

aaronjorbin commented on lib/experimental/synchronization.php:
```diff
@@ -22,3 +22,145 @@ function gutenberg_rest_api_init_collaborative_editing() {
+ * the HTML content.
+ *
+ * Legacy clients may manipulate the HTML state without updating the Yjs
+ * document. Ideally, they leave the y:gutenberg comment alone. Onca e
```
```suggestion
 * document. Ideally, they leave the y:gutenberg comment alone. Once a
```

aaronjorbin commented on lib/experimental/synchronization.php:
```diff
@@ -22,3 +22,145 @@ function gutenberg_rest_api_init_collaborative_editing() {
+	}
+	$content = stripslashes($data['post_content']);
+	// generate a new clientid for updated content that can be represented as a 53bit unsigned integer (max clientid in Yjs)
+	$ynewclientid = rand(0, 9007199254740991); // $match[4];
```
```suggestion
	$ynewclientid = wp_rand(0, 9007199254740991); // This is 2^53 – 1 which is `Number.MAX_SAFE_INTEGER` from JavaScript
```

`wp_rand`  should be a bit more random.  Also adjusted the comment so it's not unused code but instead explains the number used.

aaronjorbin commented on lib/experimental/synchronization.php:
```diff
@@ -22,3 +22,145 @@ function gutenberg_rest_api_init_collaborative_editing() {
+}
+
+// This filter must be run after all other filters! Use the highest priority number systemwide.
+add_filter( 'wp_insert_post_data', 'filter_post_content_ydoc', 20130220, 3);
```
If the goal is that it runs after all other filters, `PHP_MAX_INT` should be used.  That won't for sure make it run at the end though since it's possible for other plugins to also use it.  

Why does this need to run after everything? What happens if it doesn't?

dmonad replied:
Initially, I implemented a more complex solution that handles the edge cases that I described above. That approach required that the filter us run last so we get the latest updates.

The current approach can run anytime. I set the priority to `10` and removed the comment.

aaronjorbin commented on lib/experimental/synchronization.php:
```diff
@@ -22,3 +22,145 @@ function gutenberg_rest_api_init_collaborative_editing() {
+					$postcontent = stripslashes($post->post_content);
+					// @todo the following regex doesn't catch the ygutenberg comment
+					// preg_match('/<!-- y:gutenberg version="([a-zA-Z0-9]*)" state="([a-zA-Z0-9+\/]*={0,3})" new-content-clientid="([0-9]*)" -->/', $postcontent, $yinfo);
+					preg_match('/<!-- y:gutenberg version=\"(.*)\" state=\"(.*)\" new-content-clientid=\"(.*)\" -->/', $postcontent, $yinfo);
```
It looks like this is used both here and in `filter_post_content_ydoc`.  I wonder if this should be abstracted into it's own `get_yinfo` function that can then have automated test to ensure it correctly handles any necessary edge cases.

dmonad replied:
I fixed the issue with the proper regex and abstracted it into a `gutenberg_get_yinfo` function.

Where would I locate tests for the `sync` package? I'm pretty sure that this would be easy to implement.

aaronjorbin reviewed (commented):
This is a great start, thank you! A couple of questions and small suggestions on my first read-through.

I would like to see some automated tests before this gets too far, especially around places where others will extend this.

dmonad committed:
[Yjs Collab] Fix typo in lib/experimental/synchronization.php

Co-authored-by: Aaron Jorbin <aaronjorbin@users.noreply.github.com>

dmonad committed:
[Yjs Collab] user wp_rand instead of rand

Co-authored-by: Aaron Jorbin <aaronjorbin@users.noreply.github.com>

dmonad committed:
[Yjs Collab] Update regexes and priorities for synchronization filters

As @aaronjorbin rightfully pointed out, the new filters approach don't
need to run at the end anymore.

dmonad committed:
[Yjs Collab] implement get_yinfo function in synchronization.php

dmonad committed:
[Yjs Collab] fix tests

dmonad committed:
[Yjs Collab] fix php lint issues

dmonad committed:
[Yjs Collab] fix mobile tests

dmonad committed:
Merge branch 'trunk' into y

dmonad commented:
I fixed all tests but the E2E Tests. There is a timeout issue and I can't figure out how to fix it. I would greatly appreciate help on this.

I also wanted to implement an integration test for a collaborative session. We would need to enable the experimental feature in the e2e test and then perform some actions on two different peers. But I would also need help to implement this. But I think proper testing for the `sync` package involves a lot more work.
